libraries/chain/controller.cpp

```diff
@@ -3160,8 +3161,47 @@ struct controller_impl {
          };
          apply_s<void>(chain_head, process_new_proposer_policy);
 
+         std::optional<finalizer_policy> new_finalizer_policy;
+
+         // Process set_finalizers during transition. Only Savanna Genesis block allows it;
+         // other transition blocks ignore set_finalizers
+         auto process_new_finalizer_policy_legacy = [&](building_block::building_block_legacy& bb_legacy) -> void {
+            const auto& gpo = db.get<global_property_object>();
+            if (gpo.proposed_fin_pol_block_num.has_value() && !bb_legacy.pending_block_header_state.is_if_transition_block()) {
+               new_finalizer_policy = gpo.proposed_fin_pol;
+               new_finalizer_policy->generation = 1; // only allowed to be set once in legacy mode
+
+               db.modify( gpo, [&]( auto& gp ) {
+                  gp.proposed_fin_pol_block_num = std::nullopt;
+                  gp.proposed_fin_pol.generation = 0;
+                  gp.proposed_fin_pol.finalizers.clear();
+               });
+            }
+         };
+
+         // Process set_finalizers after transition to Savanna
+         auto process_new_finalizer_policy_savanna = [&](building_block::building_block_if& bb_savanna) -> void {
+            const auto& gpo = db.get<global_property_object>();
+            if (gpo.proposed_fin_pol_block_num.has_value()) {
+               new_finalizer_policy = gpo.proposed_fin_pol;
+               new_finalizer_policy->generation = bb_savanna.parent.finalizer_policy_generation + 1;
+
+               db.modify( gpo, [&]( auto& gp ) {
+                  gp.proposed_fin_pol_block_num = std::nullopt;
+                  gp.proposed_fin_pol.generation = 0;
+                  gp.proposed_fin_pol.finalizers.clear();
+               });
+            }
+         };
+
+         bb.apply<void>(overloaded{ process_new_finalizer_policy_legacy,
+                                    process_new_finalizer_policy_savanna });
+
```

Comment:
I don't think we need all these changes.
Just extract for both legacy and Savanna:

```
if (gpo.proposed_fin_pol_block_num.has_value()) {
              new_finalizer_policy = gpo.proposed_fin_pol;ew_finalizer_policy = gpo.proposed_fin_pol;
              db.modify( gpo, [&]( auto& gp ) {
                  gp.proposed_fin_pol_block_num = std::nullopt;
                  gp.proposed_fin_pol.generation = 0;
                  gp.proposed_fin_pol.finalizers.clear();
               });
}
```

and then call the same `assemble_block` as before (it will use the updated `new_finalizer_policy` from `building_block_common.
Even better, remove  `new_finalizer_policy` from `building_block_common, as pass the `new_finalizer_policy` as an argument to `assemble_block` (like we do for `new_proposer_policy`).
 then use the same code as before

Reply:
I will make gpo extraction code common but we still need to do differently for Legacy and Savanna to sett`generation`.

`new_finalizer_policy` was already removed from `building_block_common` and passed to `bb.assemble_block` directly by commit https://github.com/AntelopeIO/spring/pull/99/commits/7cf7cf0ea17e2f0e7fd0ce36ce1d7180dc8adc68